src/audio/mixer.c

```diff
@@ -377,6 +402,33 @@ static int mixer_copy(struct comp_dev *dev)
 		sources[i] = buffer_release(sources[i]);
 	}
 
+#if CONFIG_IPC_MAJOR_3
+	if (frames == 0 && md->sources_inactive) {
+		/*
+		 * Transition out from sources_inactive state, but
+		 * as frames==0, sources have not yet started providing
+		 * data, so push out silence until at least one source
+		 * has data. This is needed to keep downstream DAIs
+		 * fed with data and avoid XRUNs.
+		 */
+
+		frames = audio_stream_get_free_frames(&sink->stream);
```

Comment:
@singalsu This is a good question. The existing code to output silence uses this same logic, i.e. it will as much silence as the sink can fit. I used the same approach here. While running sof-test that trigger this path, it's still usually 48 frames this routine write. So we'd write more if we are really running behind and then it probably makes sense to write more silence to avoid a DAI XRUNs. When the upstream components (that provide data to mixer) are started, there is probably risk of some extra delays in the initial copy() iterations of those components (shouldn't happen but the risk is still there), so definitely not a good time to run low on buffering towards the DAI(s).